File: trunk/TAO/orbsvcs/orbsvcs/Trader/Property_Evaluator.cpp
// File: Property_Evaluator.cpp
// $Id$

#include "Trader.h"
#include "Property_Evaluator.h"

// Constructor

TAO_Property_Evaluator::
TAO_Property_Evaluator(const CosTrading::PropertySeq& props,
		       CORBA::Boolean supports_dp)
  : props_ (props),
    supports_dp_ (supports_dp)
{
}


TAO_Property_Evaluator::
TAO_Property_Evaluator(CosTrading::Offer& offer,
		       CORBA::Boolean supports_dp)
  : props_ (offer.properties),
    supports_dp_ (supports_dp)
{
}

int
TAO_Property_Evaluator::is_dynamic_property (int index)
{
  CORBA::Environment env;
  int return_value = 0,
    num_properties = this->props_.length();

  // Ensure index is in bounds. 
  if (index >= 0 && index < num_properties)
    {
      // Obtain the value of the property at index <index>.
      const CORBA::Any& value = this->props_[index].value;
      CORBA::TypeCode* type = value.type();
      
      if (type->equal(CosTradingDynamic::_tc_DynamicProp, env))
	return_value = 1;
    }
  
  return return_value;
}

CORBA::Any*
TAO_Property_Evaluator::property_value(int index,
				       CORBA::Environment& _env)
  TAO_THROW_SPEC ((CosTradingDynamic::DPEvalFailure))
{
  CORBA::Any* prop_val = 0;
  
  if (! this->is_dynamic_property(index))
    prop_val = (CORBA::Any *) &(this->props_[index].value);
  else if (this->supports_dp_)
    {
      // Property is defined at this point.
      DP_Eval* dp_eval;
      CosTradingDynamic::DynamicProp* dp_struct;
      const CORBA::String_var name = this->props_[index].name;
      const CORBA::Any& value = this->props_[index].value;

      // Extract the DP_Struct.
      value >>= dp_struct;
      dp_eval = dp_struct->eval_if;
	  
      if (CORBA::is_nil (dp_eval))
	TAO_THROW_RETURN (CosTradingDynamic::DPEvalFailure (), prop_val);
      else
	{
	  CORBA::TypeCode* type = dp_struct->returned_type;
	  CORBA::Any& info = dp_struct->extra_info;
	  
	  TAO_TRY
	    {
	      // Retrieve the value of the dynamic property.
	      prop_val = dp_eval->evalDP(name, type, info, TAO_TRY_ENV);
	    }
	  TAO_CATCH (CORBA::SystemException, excp)
	    {
	      TAO_THROW_RETURN
		(CosTradingDynamic::DPEvalFailure (name, type, info),
		 prop_val);
	    }
	  TAO_ENDTRY;
	}
    }

  return prop_val;
}

CORBA::TypeCode*
TAO_Property_Evaluator::property_type(int index)
{
  CORBA::TypeCode* prop_type = CORBA::TypeCode::_nil();
  
  // Determine if property is both defined and dynamic.
  if (this->is_dynamic_property (index))
    {
      // Extract type information from the DP_Struct.
      const CORBA::Any& value = this->props_[index].value;
      CosTradingDynamic::DynamicProp* dp_struct;
      value >>= dp_struct;
      
      // Grab a pointer to the returned_type description
      prop_type = CORBA::TypeCode::_duplicate(dp_struct->returned_type);
    }
  else
    // TypeCode is self-evident at this point.
    prop_type = this->props_[index].value.type();
  
  return prop_type;
}

TAO_Property_Evaluator_By_Name::
TAO_Property_Evaluator_By_Name (const CosTrading::PropertySeq& properties,
				CORBA::Environment& _env,
				CORBA::Boolean supports_dp)
  TAO_THROW_SPEC ((CosTrading::DuplicatePropertyName,
		   CosTrading::IllegalPropertyName))
    : TAO_Property_Evaluator (properties, supports_dp)
{
  string prop_name;
  int length = this->props_.length();

  for (int i = 0; i < length; i++)
    {
      const CosTrading::Property& prop = this->props_[i];

      if (! TAO_Trader_Base::is_valid_identifier_name (prop.name))
	TAO_THROW (CosTrading::IllegalPropertyName (prop.name));
      
      prop_name = prop.name;
      if (! (this->table_.insert (make_pair (prop_name, i))).second)
	TAO_THROW (CosTrading::DuplicatePropertyName (prop.name));
    }
}

TAO_Property_Evaluator_By_Name::
TAO_Property_Evaluator_By_Name(CosTrading::Offer& offer,
			       CORBA::Boolean supports_dp)
  : TAO_Property_Evaluator(offer, supports_dp)
{
  int length = this->props_.length();

  for (int i = 0; i < length; i++)
    {
      string prop_name = (const char*) this->props_[i].name;
      this->table_[prop_name] = i;
    }
}

int
TAO_Property_Evaluator_By_Name::
is_dynamic_property(const char* property_name)
{
  int predicate = 0;
  string prop_name (property_name);
  Lookup_Table_Iter lookup_iter = this->table_.find (prop_name);

  // If the property name is in the map, delegate evaluation to our
  // superclass. Otherwise, throw an exception.
  if (lookup_iter != this->table_.end())
    {
      int index = (*lookup_iter).second;
      predicate = TAO_Property_Evaluator::is_dynamic_property(index);
    }

  return predicate;
}
  
CORBA::Any*
TAO_Property_Evaluator_By_Name::property_value(const char* property_name,
					       CORBA::Environment& _env)
  TAO_THROW_SPEC ((CosTradingDynamic::DPEvalFailure))
{
  CORBA::Any* prop_value = 0;
  string prop_name(property_name);
  Lookup_Table_Iter lookup_iter = this->table_.find(prop_name);

  // If the property name is in the map, delegate evaluation to our
  // superclass. Otherwise, throw an exception.
  if (lookup_iter != this->table_.end())
    {
      int index = (*lookup_iter).second;
      prop_value = TAO_Property_Evaluator::property_value(index, _env);
    }

  return prop_value;
}
  
CORBA::TypeCode*
TAO_Property_Evaluator_By_Name::property_type(const char* property_name)
{
  string prop_name(property_name);
  CORBA::TypeCode* prop_type = CORBA::TypeCode::_nil();
  Lookup_Table_Iter lookup_iter = this->table_.find(prop_name);
  
  // If the property name is in the map, delegate evaluation to our
  // superclass. Otherwise, throw an exception.
  if (lookup_iter != this->table_.end())
    {
      int index = (*lookup_iter).second;
      prop_type = TAO_Property_Evaluator::property_type(index);
    }

  return prop_type;
}

const CosTrading::Property*
TAO_Property_Evaluator_By_Name::get_property (const char* property_name)
{
  CosTrading::Property* property = 0;
  string prop_name (property_name);
  Lookup_Table_Iter lookup_iter = this->table_.find(prop_name);

  if (lookup_iter != this->table_.end())
    property = (CosTrading::Property *) &this->props_[(*lookup_iter).second];

  return property;
}

File: tags/TAO-0_1_38/TAO/orbsvcs/tests/AVStreams/server_discovery/Property_Evaluator.java
/* -*- C++ -*- */

// ========================================================================
// $Id$
// 
// = FILENAME
//    Property_Evaluator.java
// 
// = AUTHOR 
//    Seth Widoff <sbw1@cs.wustl.edu> 
//   
// ========================================================================


import org.omg.CORBA.*;
import CosTrading.*;
import CosTradingDynamic.*;
import TAO_Sequences.*;
import TAO_VR.*;

/**
  * This is a Java port of the TAO_Property_Evaluator class in the TAO
  * trading service implementation.
  */

public class Property_Evaluator
{
  private Property[] props_;
  private boolean supports_dp_ = true;
  
  /**
    * Construct an instance of TAO_Property_Evaluator that operates on
    * a PropertySeq, <properties>.
    */
  Property_Evaluator (Property[] properties)
    {
      this.props_ = properties;
    }

  /**
    * Construct an instance of TAO_Property_Evaluator that operates on
    * an <offer>.
    */
  public Property_Evaluator (Offer offer)
    {
      this.props_ = offer.properties;
    }

  /**
    * Disable or enable the evaluation of dynamic properties.
    */
  public void allow_dynamic_properties (boolean toggle)
    {
      this.supports_dp_ = toggle;
    }

  /**
    * Returns true if the property at index <index> is dynamic. Returns a
    * 0 when the index is out of bounds.    
    */
  public boolean is_dynamic_property (int index)
    {
      boolean return_value = false;
      int num_properties = this.props_.length;
      
      if (index >= 0 && index < num_properties)
	return_value = Property_Evaluator.is_dynamic_property (this.props_[index]);
      
      return return_value;
    }	 

  public static boolean is_dynamic_property (Property prop)
    {
      boolean return_value = false;

      Any value = prop.value;
      TypeCode type = value.type ();

      // Compare the type of the Any contents with the typecode
      // for a dynamic property structure. 
      if (type.equal (DynamicPropHelper.type ()))
	return_value = true;

      return return_value;
    }
  
  /**
    * Returns value of the property whose index is <index>. If the
    * property at that index is dynamic and the trader supports dynamic 
    * properties, then the property_value method will obtain the value
    * of the dynamic property using the evalDP method on the
    * CosTradingDynamic::DynamicPropEval interface, passing on a
    * CosTradingDynamic::DPEvalFailure exception on failure. If the
    * property index is undefined, the method returns a null pointer.
    */
  public Any property_value (int index, ORB orb)
    throws DPEvalFailure
    {
      Any return_value = null;
      int num_properties = this.props_.length;
      
      if (index >= 0 && index < num_properties)
	{
	  return_value =
	    Property_Evaluator.property_value (this.props_[index], orb);
	}

      return return_value;
    }

  public static Any property_value (Property prop, ORB orb)
    throws DPEvalFailure
    {
      Any prop_value = null;
      
      if (! Property_Evaluator.is_dynamic_property (prop))
	prop_value = prop.value;
      else
	{
	  DynamicProp dp_struct =
	    DynamicPropHelper.extract (prop.value);
	  String prop_name = prop.name;
	  
	  // This is because of the bug in TAO:
	  // Now we extract the object reference from the dynamic
	  // property structure.
	  org.omg.CORBA.Object obj =
	    orb.string_to_object (dp_struct.eval_if);
	  DynamicPropEval dp_eval = DynamicPropEvalHelper.narrow (obj);

	  if (dp_eval != null)
	    {
	      prop_value = dp_eval.evalDP (prop_name,
					   dp_struct.returned_type,
					   dp_struct.extra_info);
	    }
	  else
	    {
	      throw new DPEvalFailure (prop_name,
				       dp_struct.returned_type,
				       dp_struct.extra_info);	      
	    }
	}

      return prop_value;
    }
  
  /**
    * Returns the type of the property whose index is <index>. If the
    * property is dynamic and the trader supports dynamic properties,
    * then the method returns the <returned_type> field of the
    * CosTradingDynamic::DynamicProp struct associated with the
    * property name. If the index is out of bounds, the method returns 
    * a null pointer (that is, 0).
    */
  
  public TypeCode property_type (int index)
    {
      TypeCode prop_type = null;

      if (! this.is_dynamic_property (index))
	prop_type = this.props_[index].value.type ();
      else
	{
	  Any value = this.props_[index].value;
	  DynamicProp dp_struct = DynamicPropHelper.extract (value);

	  prop_type = dp_struct.returned_type;
	}

      return prop_type;
    }

  public String property_to_string (int index, ORB orb)
    throws DPEvalFailure
    {
      String return_value = null;
      Any any_value = this.property_value (index, orb);
      TypeCode type = any_value.type ();
      int kind = type.kind ().value ();
      
      if (kind == TCKind._tk_null)
	return_value = "Null";

      else if (kind == TCKind._tk_void)
	return_value = "Void";
      
      else if (kind == TCKind._tk_short)
	return_value = String.valueOf (any_value.extract_short ());

      else if (kind == TCKind._tk_long)
	return_value = String.valueOf (any_value.extract_long ());

      else if (kind == TCKind._tk_ushort)
	return_value = String.valueOf (any_value.extract_ushort ());

      else if (kind == TCKind._tk_ulong)
	return_value = String.valueOf (any_value.extract_ulong ());

      else if (kind == TCKind._tk_float)
	return_value = String.valueOf (any_value.extract_float ());

      else if (kind == TCKind._tk_double)
	return_value = String.valueOf (any_value.extract_double ());

      else if (kind == TCKind._tk_boolean)
	return_value = String.valueOf (any_value.extract_boolean ());

      else if (kind  == TCKind._tk_char)
	return_value = String.valueOf (any_value.extract_char ());

      else if (kind == TCKind._tk_string)
	return_value = any_value.extract_string ();

      else if (type.equal (StringSeqHelper.type ()))
	{
	  System.out.println ("Printing string sequence.");
	  String[] str_seq = StringSeqHelper.extract (any_value);	  

	  for (int length = str_seq.length, i = 0; i < length; i++)
	    return_value = str_seq[i] + " ";	 
	}

      else if (type.equal (ULongSeqHelper.type ()))
	{
	  System.out.println ("Printing ulong sequence.");
	  int[] ulong_seq = ULongSeqHelper.extract (any_value);	  

	  for (int length = ulong_seq.length, i = 0; i < length; i++)
	    return_value = ulong_seq[i] + " ";
	}

      else
	System.out.println ("TCKind" + kind);

      return return_value;
    }
}
